File: backend/market_platform/schemas/ai.py
from typing import Any, Literal

from pydantic import BaseModel, Field

AiIntent = Literal["screen", "compare", "explain_stock", "rank", "unknown"]
AiSortDir = Literal["asc", "desc"]


class AiQueryRequest(BaseModel):
    query: str = Field(min_length=3, max_length=500)
    exchange: str = "NSE"
    limit: int = Field(default=25, ge=1, le=100)
    include_explanation: bool = True


class AiQueryFilter(BaseModel):
    field: str
    operator: str
    value: str | int | float | bool | list[Any] | None = None


class AiQueryRow(BaseModel):
    symbol: str
    company_name: str | None = None
    sector: str | None = None
    industry: str | None = None
    ltp: float | None = None
    market_cap: float | None = None
    pe: float | None = None
    trend_score: int | None = None
    trend: str | None = None
    momentum_score: float | None = None
    momentum_acceleration: float | None = None
    return_1m: float | None = None
    return_3m: float | None = None
    return_6m: float | None = None
    volume_ratio: float | None = None
    distance_from_52w_high: float | None = None


class AiQueryResponse(BaseModel):
    answer: str
    interpreted_query: str
    filters: list[AiQueryFilter]
    rows: list[AiQueryRow]
    total: int
    warnings: list[str] = Field(default_factory=list)


class AiStatusResponse(BaseModel):
    enabled: bool
    provider: str | None = None
    model: str | None = None


class SafeAiQueryPlan(BaseModel):
    intent: AiIntent = "screen"
    symbols: list[str] = Field(default_factory=list)
    sector: str | None = None
    industry: str | None = None
    filters: list[AiQueryFilter] = Field(default_factory=list)
    sort_by: str = "momentum_score"
    sort_dir: AiSortDir = "desc"
    limit: int = Field(default=25, ge=1, le=100)

File: web/src/types/ai.ts
export type AiQueryFilter = {
  field: string;
  operator: string;
  value: string | number | boolean | null | Array<string | number | boolean>;
};

export type AiQueryRow = {
  symbol: string;
  company_name?: string | null;
  sector?: string | null;
  industry?: string | null;
  ltp?: number | null;
  market_cap?: number | null;
  pe?: number | null;
  trend_score?: number | null;
  trend?: string | null;
  momentum_score?: number | null;
  momentum_acceleration?: number | null;
  return_1m?: number | null;
  return_3m?: number | null;
  return_6m?: number | null;
  volume_ratio?: number | null;
  distance_from_52w_high?: number | null;
};

export type AiQueryResponse = {
  answer: string;
  interpreted_query: string;
  filters: AiQueryFilter[];
  rows: AiQueryRow[];
  total: number;
  warnings: string[];
};

export type AiStatusResponse = {
  enabled: boolean;
  provider?: string | null;
  model?: string | null;
};
